voila/app.py

```diff
@@ -324,267 +269,120 @@ def _default_root_dir(self):
-        self.log.debug('using template: %s', self.voila_configuration.template)
-        self.log.debug('nbconvert template paths:\n\t%s', '\n\t'.join(self.nbconvert_template_paths))
-        self.log.debug('template paths:\n\t%s', '\n\t'.join(self.template_paths))
-        self.log.debug('static paths:\n\t%s', '\n\t'.join(self.static_paths))
-        if self.notebook_path and not os.path.exists(self.notebook_path):
-            raise ValueError('Notebook not found: %s' % self.notebook_path)
-
-    def _handle_signal_stop(self, sig, frame):
-        self.log.info('Handle signal %s.' % sig)
-        self.ioloop.add_callback_from_signal(self.ioloop.stop)
-
-    def start(self):
-        self.connection_dir = tempfile.mkdtemp(
-            prefix='voila_',
-            dir=self.connection_dir_root
-        )
-        self.log.info('Storing connection files in %s.' % self.connection_dir)
-        self.log.info('Serving static files from %s.' % self.static_root)
+                self.template,
+            )
 
-        self.kernel_spec_manager = KernelSpecManager(
-            parent=self
-        )
-
-        self.kernel_manager = MappingKernelManager(
-            parent=self,
-            connection_dir=self.connection_dir,
-            kernel_spec_manager=self.kernel_spec_manager,
-            allowed_message_types=[
-                'comm_open',
-                'comm_close',
-                'comm_msg',
-                'comm_info_request',
-                'kernel_info_request',
-                'shutdown_request'
-            ]
-        )
+        template_conf_dir = [os.path.join(k, '..') for k in self.nbconvert_template_paths]
+        conf_paths = [os.path.join(d, 'conf.json') for d in template_conf_dir]
+        for p in conf_paths:
+            # see if config file exists
+            if os.path.exists(p):
+                # load the template-related config
+                with open(p) as json_file:
+                    conf = json.load(json_file)
+                # update the overall config with it, preserving CLI config priority
+                if 'traitlet_configuration' in conf:
+                    recursive_update(conf['traitlet_configuration'], self.config.Voila)
+                    # pass merged config to overall voila config
+                    self.extension_config.Voila = Config(conf['traitlet_configuration'])
+        self.serverapp.log.debug("using template: %s", self.template)
+        self.serverapp.log.debug("nbconvert template paths:\n\t%s", "\n\t".join(self.nbconvert_template_paths))
+        self.serverapp.log.debug("template paths:\n\t%s", "\n\t".join(self.template_paths))
+        self.serverapp.log.debug("static paths:\n\t%s", "\n\t".join(self.static_paths))
 
         jenv_opt = {"autoescape": True}  # we might want extra options via cmd line like notebook server
         env = jinja2.Environment(loader=jinja2.FileSystemLoader(self.template_paths), extensions=['jinja2.ext.i18n'], **jenv_opt)
```

Comment:
I wonder if we could make use of `ExtensionAppJinjaMixin` here: https://github.com/jupyter/jupyter_server/blob/5f99e53f7206c1272494a977e3246aa870d74fdf/jupyter_server/extension/application.py#L89